src/pybind/mgr/ssh/module.py

```diff
@@ -0,0 +1,725 @@
+        return list(map(lambda r: r.get(), self._result))
+
+    @property
+    def is_complete(self):
+        return all(map(lambda r: r.ready(), self._result))
+
+class SSHReadCompletionReady(SSHReadCompletion):
+    def __init__(self, result):
+        self._result = result
+
+    @property
+    def result(self):
+        return self._result
+
+    @property
+    def is_complete(self):
+        return True
+
+class SSHWriteCompletion(orchestrator.WriteCompletion):
+    def __init__(self, result):
+        if isinstance(result, multiprocessing.pool.AsyncResult):
+            self._result = [result]
+        else:
+            self._result = result
+        assert isinstance(self._result, list)
+
+    @property
+    def result(self):
+        return list(map(lambda r: r.get(), self._result))
+
+    @property
+    def is_persistent(self):
+        return all(map(lambda r: r.ready(), self._result))
+
+    @property
+    def is_effective(self):
+        return all(map(lambda r: r.ready(), self._result))
+
+    @property
+    def is_errored(self):
+        for r in self._result:
+            if not r.ready():
+                return False
+            if not r.successful():
+                return True
+        return False
+
+class SSHWriteCompletionReady(SSHWriteCompletion):
+    def __init__(self, result):
+        self._result = result
+
+    @property
+    def result(self):
+        return self._result
+
+    @property
+    def is_persistent(self):
+        return True
+
+    @property
+    def is_effective(self):
+        return True
+
+    @property
+    def is_errored(self):
+        return False
+
+class SSHOrchestrator(MgrModule, orchestrator.Orchestrator):
+
+    _STORE_HOST_PREFIX = "host"
+    _DEFAULT_INVENTORY_CACHE_TIMEOUT_MIN = 10
+
+    MODULE_OPTIONS = [
+        {'name': 'ssh_config'},
+        {'name': 'inventory_cache_timeout_min'},
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(SSHOrchestrator, self).__init__(*args, **kwargs)
+        self._cluster_fsid = None
+        self._worker_pool = multiprocessing.pool.ThreadPool(1)
+
+    @staticmethod
+    def can_run():
+        if remoto is not None:
+            return True, ""
+        else:
+            return False, "loading remoto library:{}".format(
+                    remoto_import_error)
+
+    def available(self):
+        """
+        The SSH orchestrator is always available.
+        """
+        return self.can_run()
+
+    def wait(self, completions):
+        self.log.info("wait: completions={}".format(completions))
+
+        complete = True
+        for c in completions:
+            if not isinstance(c, SSHReadCompletion) and \
+                    not isinstance(c, SSHWriteCompletion):
+                raise TypeError("unexpected completion: {}".format(c.__class__))
+
+            if c.is_complete:
+                continue
+
+            complete = False
+
+        return complete
+
+    @staticmethod
+    def time_from_string(timestr):
```

Comment:
This duplicates `crash.module.Module#time_from_string` and `telemetry.module.Module#parse_timestamp`. Does it make sense to movie this helper to `mgr_util.py`?

Reply:
i don't think it makes sense. e.g. in the crash module the format of the timestamp is also tied to an external json schema.